### ISSUE TITLE
I've fixed the sluggish whole-page-redraw-on-issue-drag!

### ISSUE THREAD

**Issue by ChangePlaces** [kanboard/kanboard#4197]
### Actual behaviour

When a project has several swimlanes with a few hundred issues across them, dragging one issue locks the page for 5-10+ seconds, relative to the amount of issues there are.

### Expected behaviour

A quick drag-and-drop.

- Kanboard version: all versions
- Database type and version: sqlite
- PHP version: 7.3

I'm invested in this now and would like to help. Why does the system redraw the whole page on a drag-and-drop, why is it necessary, and as-I-suspect was there a gotcha that caused you to take this action? 

UPDATE:

## "The Problem"

When a task is dragged and dropped to a different position than its start position, it triggers a request to the server to save the new position with the task id and this causes the server to return the whole board of tasks and then redraws all the tasks. This fetching of data (for me is currently 2.5MB of data and) is the cause of the delay in processing (fetching takes about 5 seconds, redrawing takes about 2s), and needs to be eliminated. Some redrawing does need to take place to update the position data of the task that is moved and also the data of the other tasks in the column the task was moved from and to.

## The solution(s)

There are two possible solutions, one ideal and ~~hard to implement~~ quite easy to implement, one not so ideal, but ~~easier~~ harder and uglier to implement:

1. The task that is dragged has its position data updated on its drop (swimlane, column, position in the col), and the other affected tasks also have their position data updated locally. This will involve no data being returned from the server and is the fastest solution. Changing the dropped task data is easy as it's known, however, the source and destination columns will have to have its tasks cycled through and increase the position number (starting at 1). 

Actually now that I break it down, this isn't as hard as I thought it may be, and will be the solution I will tackle!

2. The server returns the tasks of the two affected cols and locally redrawn. This is probably overkill as I may be able to swing solution 1 and will probably be harder to implement anyway!

**Comment by ChangePlaces** [comment]
OK, this is fixed! A simple to solution 1. There is a pause for about... 0.5s... but I think that's acceptable :)

I've had to change the app/Controller/BoardAjaxController and app.min.js. I've had to unminify the app.min.js, so what minifier do you use and I'll get it compressed up and sent as a PR!

Though, I'd like to recommend that you include an unminified version for people to look at to make alterations.

**Comment by rafacamargo123** [comment]
Nice job!
About the unminified assets, they already exists on git repo, inside assets folder. 
See more about it here: https://docs.kanboard.org/en/latest/developer_guide/assets.html

**Comment by ChangePlaces** [comment]
Nice, thanks. I had a thought over the night and wanted to ask if there were other reasons for the whole page refresh when dragging an issue? Maybe because of triggers modifying issues or something? I don't make full use of this (I use it as a basic kanboard to log issues with projects and swimlanes) so what works for me may not work in the bigger picture!

**Comment by brblair** [comment]
What were the code changes you made?  Are you willing to share your updates?  Thank you.

**Comment by ChangePlaces** [comment]
Actually, I haven't uploaded them because it's far bigger than I thought! It turns out that some architecture changes need to be made as with my changes, the layout does redraw columns after a task drag. It's been a while now, but it was a finicky few changes.

**Comment by nekohayo** [comment]
Hi @ChangePlaces, maybe you can share your WIP in a branch somewhere so that others (if any) can look and help (if you need some help, that is). I'm simply a user here but I would like to thank you in advance and encourage you to see this through the end; the performance limitations of Kanboard are the one thing scaring me from using it for anything serious. The good news is that it seems you have identified an area where Kanboard's code does "too much work", and I know that needless work represent 80-90% of performance problems in software engineering (micro-optimizations are rarely the cause).

If it helps, in order to benchmark various hosts/setups, I've generated a kanboard-importable .csv file with about 1300 tasks spread across 10 columns (but only 1 swimlane), which results in a 2MB sqlite file. I could share that with anyone needing an easy "stress test", though I'm wondering if that is sufficient as a true "torture test" (vs something that has 10 000 tasks, for example). But for what it's worth, even with 1300 tasks, it brings a shared host service to its knees; trying to display the board works at ~500 tasks, but at 1300 it doesn't even try, it chokes and dies with a "`Internal Error: SQL Error: SQLSTATE[HY000]: General error: 1 too many SQL variables`" (not sure if that's a host limitation or if that would be a kanboard bug in the way it interacts with sqlite)

**Comment by ChangePlaces** [comment]
As a single-dev I've had to put kanboard to one side, and I've taken delight in using an A5 exercise book for tasks. It's far more handy than I thought it would be and so more free-ing than from a computer!

**Comment by nekohayo** [comment]
For me and many others out there the notebook/whiteboard technique cannot scale. I have dozens of projects and manage teams, which means keeping track of hundreds (if not thousands) of tasks... and I can't imagine myself throwing those out of the window and just using a "short" todo list. In fact, I've already got a todo list, and I'm hoping to migrate my workflow from that to Kanboard (or at least the projects-style portions of it)... but I need to be sure that Kanboard's UI can scale to handle that amount of items.

I'm sure your performance improvements would be appreciated by Kanboard users. How far along were you, does it really need to be "perfect" or were your optimizations already working (without breaking things) and "step of improvement" that can be delivered? If so why not send in a merge request?

**Comment by rafacamargo123** [comment]
I work on a team using the same project on kanboard for over 3 years. We have over 5000 tasks on a single board. Of course, they are not all open. We close the “done” tasks every week. We don’t have problems searching for tasks or using kanboard, as long as the search doesn’t return thousands of tasks. What happens is that the browser freezes for some seconds if I clear the filter completely, but I think it’s because the JavaScript running on thousands of dom elements. We don’t have performance issues searching for tasks or using kanboard. It just doesn’t make sense to keep all tasks ever always opened and showing on the board. 
On our company, each team has a board, and for bigger projects we create a board for every project

**Comment by nekohayo** [comment]
@rafacamargo123 doesn't it crash for you if you go view the board's closed tasks (without searching)? Even though you can "workaround" with search, it's still a bit scary to have such a bug occur in the first place.

In any case there's no reason why that bug should be left to exist, in theory; I've used applications that deal with sqlite databases over 100 MB in size (that's a _lot_ of records) and they manage fine. This issue here only occurs in some views, like the "board" view.

**Comment by rafacamargo123** [comment]
No, the kanboard itself does not crash, the browser that does slow down. 
About this “too many variables” error, while developing a plugin this week I found out how to add subqueries to where statements with piccodb and I think it will be quite simple to resolve the issue. 
When I have more free time, maybe during the weekend, I think I will be able to close that particular issue.
I don’t experience this issue because we use kanboard with MySQL. 
We just got to the task # 10000 today on our kanboard with 3 big active projects, others with <100 tasks!
